src/net.cpp

```diff
@@ -127,6 +127,12 @@ size_t CSerializedNetMsg::GetMemoryUsage() const noexcept
     return sizeof(*this) + memusage::DynamicUsage(data);
 }
 
+size_t CNetMessage::GetMemoryUsage() const noexcept
+{
+    // Don't count the dynamic memory used for the m_type string (see CSerializedNetMsg::GetMemoryUsage),
```

Comment:
The comment in the referred function explains it.
```
    // Don't count the dynamic memory used for the m_type string, by assuming it fits in the
    // "small string" optimization area (which stores data inside the object itself, up to some
    // size; 15 bytes in modern libstdc++).
```
 It has to do with the internal implementation of std::string, where it has a fixed-size buffer that it will use for short strings (such as the message types) before allocating memory on the heap.

Reply:
i do agree that over-counting is probably better than under-counting here, but i just took this over from `CSerializedNetMsg` as i assumed this was the right thing to do. If we change it it needs to be changed in both places.

Reply:
For g++ 13.2 on x86_64 it checks out, this returns 16:
```c++
#include <string>
#include <iostream>

bool data_is_internal(const std::string &s)
{
    return s.data() >= reinterpret_cast<const char*>(&s) && s.data() + s.size() - sizeof(s) <= reinterpret_cast<const char*>(&s);
}

int main()
{
    std::string s;
    while (data_is_internal(s)) {
        s += 'a';
    }
    std::cout << s.size() << std::endl;
    return 0;
}
```
But yes it's not possible to guarantee across c++ compilers and platforms so adding the check to `DynamicUsage` makes sense imo.

Reply:
Alternatively, maybe a test checking this exact scenario could help:
```C++
BOOST_AUTO_TEST_CASE(sso_size_test)
{
    CSerializedNetMsg empty_msg;

    CSerializedNetMsg msg_with_type;
    msg_with_type.m_type = std::string(CMessageHeader::COMMAND_SIZE, 'X');

    BOOST_CHECK_EQUAL(sizeof(msg_with_type), sizeof(empty_msg));
}
```

Reply:
Thanks for the hints, @sipa!

According to https://en.cppreference.com/w/cpp/utility/compare/compare_three_way, this also seems safe to use:
```c++
return (msg_start <=> s.data()) <= 0 && (s.data() <=> msg_end) <= 0;
```
or
```c++
return (msg_start <=> s.data()) != std::strong_ordering::greater
      && (s.data() <=> msg_end) != std::strong_ordering::greater;
```

Reply:
> i'm kind of hestitant to add a test like that because i'm not sure we want to require this behavior from the C++ library.

I understood the memory calculations will be incorrect if this would fail - is that not the case?

Reply:
> I understood the memory calculations will be incorrect if this would fail - is that not the case?

Yes, unless we add the logic to `memusage.h`. i guess i've always seen memusage more as a "best effort estimate" than something that needs to be correct to the byte, because it's so hard to guarantee across implementations, we don't really want to extend our scope to C++ library implementation specifics. But i dunno.

BTW: It looks like currently. memusage.h doesn't handle `std::string` (nor `std::basic_string` at all. This would be new.

Reply:
Done